go/arrow/avro/schema.go

```diff
@@ -0,0 +1,409 @@
+			}
+		} else {
+			// avro "record" type, node has "fields" array
+			if node.ofType.(string) == "record" {
+				var n schemaNode
+				n.name = node.name
+				n.ofType = node.ofType
+				if len(node.fields) > 0 {
+					n.fields = append(n.fields, node.fields...)
+				}
+				f := iterateFields(n.fields)
+				return arrow.Field{Name: node.name, Type: arrow.StructOf(f...)}
+			}
+		}
+	// Avro complex types
+	case map[string]interface{}:
+		//var n schemaNode
+		//n.name = node.name
+		//n.ofType = node.ofType.(map[string]interface{})["type"]
+		switch node.logicalType {
+		case "":
+			return avroComplexToArrowField(node)
+		default:
+			return avroLogicalToArrowField(node)
+		}
+
+	// Avro union types
+	case []interface{}:
+		var unionTypes []string
+		for _, ft := range node.ofType.([]interface{}) {
+			switch ft.(type) {
+			// primitive types
+			case string:
+				if ft != "null" {
+					unionTypes = append(unionTypes, ft.(string))
+				}
+				continue
+			// complex types
+			case map[string]interface{}:
+				var n schemaNode
+				n.name = node.name
+				n.ofType = ft.(map[string]interface{})["type"]
+				if _, f := ft.(map[string]interface{})["fields"]; f {
+					for _, field := range ft.(map[string]interface{})["fields"].([]interface{}) {
+						n.fields = append(n.fields, field.(map[string]interface{}))
+					}
+				}
+				f := iterateFields(n.fields)
+				return arrow.Field{Name: node.name, Type: arrow.StructOf(f...)}
+			}
+		}
+		// Supported Avro union type is null + one other type.
+		// TODO: Complex AVRO union to Arrow Dense || Sparse Union.
+		if len(unionTypes) == 1 {
+			return arrow.Field{Name: node.name, Type: AvroPrimitiveToArrowType(unionTypes[0])}
+		} else {
+			// BYTE_ARRAY is the catchall if union type is anything beyond null + one other type.
+			return arrow.Field{Name: node.name, Type: arrow.BinaryTypes.Binary}
+		}
+	}
+	return arrow.Field{Name: node.name, Type: arrow.BinaryTypes.Binary}
+}
+
+func avroLogicalToArrowField(node schemaNode) arrow.Field {
+	// Avro logical types
+	switch node.logicalType {
+	// The decimal logical type represents an arbitrary-precision signed decimal number of the form unscaled × 10-scale.
+	// A decimal logical type annotates Avro bytes or fixed types. The byte array must contain the two’s-complement
+	// representation of the unscaled integer value in big-endian byte order. The scale is fixed, and is specified
+	// using an attribute.
+	//
+	// The following attributes are supported:
+	// scale, a JSON integer representing the scale (optional). If not specified the scale is 0.
+	// precision, a JSON integer representing the (maximum) precision of decimals stored in this type (required).
+	case "decimal":
+		if node.precision <= 38 {
+			return arrow.Field{Name: node.name, Type: &arrow.Decimal128Type{Precision: node.precision, Scale: node.scale}}
+		} else {
+			return arrow.Field{Name: node.name, Type: &arrow.Decimal256Type{Precision: node.precision, Scale: node.scale}}
+		}
```

Comment:
probably easier to simplify this function and switch to something like:

```go
var dt arrow.DataType
switch node.logicalType {
// ...
case "decimal":
     id := arrow.DECIMAL128
     if node.precision > 38 {
             id = arrow.DECIMAL256
     }
     dt, _ = arrow.NewDecimalType(id, node.precision, node.scale)
case "uuid":
     dt = types.NewUUIDType() // import "github.com/apache/arrow/v13/internal/types"
// etc.....
}
return arrow.Field{Name: node.name, Type: dt}
```

Reply:
making the change - will be in next commit